### ISSUE TITLE
[FEATURE_REQUEST] All lore, instruct, cards inserted at near end of context to utilize context shifting

### ISSUE THREAD

**Issue by Drael64** [SillyTavern/SillyTavern#1316]
### Is your feature request related to a problem? If so, please describe.

Context shifting in koboldcpp can save inference time by reproducing already existing context at the beginning of the chat. However this is never used in Silly Tavern because most of the changing info is inserted at the start. 

### Describe the solution you'd like

I stripped all the instruct and world data from the start with a blank story string, and put the instruct section in 2 replies deep insider square brackets via authors notes (and explained that was what it was, to continue the story outside the brackets) - and the model used responded just fine. 

Of course some of the variables from the story string didn't work when placed there in Authors Notes.  

But from my test, I suspect you can put all the info - the activated lore entries, the character cards, the instruct section, the whole shebang, in this location - a few entries from the end of chat/context. 

And that would allow people to always use context shifting in silly tavern, saving inference time, and possible running more powerful models or using larger context than your machine could normally handle. 

So basically the idea is - a way to shift all that stuff to a set number of replies from the end of chat, rather than at the start of chat, in order to utilized context shifting in Silly Tavern. 

### Priority

Medium (Would be very useful)

**Comment by Cohee1207** [comment]
I'm not sure I understand the exact context composition you want to achieve. As far as I understand from concedo's explanation - as long as the beginning of the context (memory in Kobold terms, story string in ST terms) stays the same as the chat progresses - this context shifting will kick in if you don't use World Info and other injections.

**Comment by Drael64** [comment]
> I'm not sure I understand the exact context composition you want to achieve. As far as I understand from concedo's explanation - as long as the beginning of the context (memory in Kobold terms, story string in ST terms) stays the same as the chat progresses - this context shifting will kick in if you don't use World Info and other injections.

So basically context shifting will keep anything towards of the start of the context window that is just normal 'scrolling' chat, so long as there are no fixed elements, or contextually changing elements there at the START of the context window. It's  very much based on WHERE in the context window those elements appear. 

So if your context window/prompt goes:

_*Raw chat stuff*
*More raw chat stuff*
*Lore Entries/Character Card/Instruction*
*Two more chat entries*_

Then you will always have context shifting working, and it will only regenerate the stuff after "More raw chat entries". 

Essentially the idea here is to have EVERYTHING operate like an Author's Note that's placed just before the end of chat. So that the shifting context never needs to regenerate that first blob of raw chat for prompt processing. 

If you format right at least -In my test I put it all in square brackets [], also prompt explained that the text there was instructions, and the text to continue with was as the 'roleplay story' that was the outside the brackets (much like the prompts for getting main model summaries in ST work) - if you format it right, when placed in that area, just before the end of chat, even Mistral seems to handle lots of instruct data of there, just fine. And Mistral is not amazing at instruct. 

If I could get this formatting then I could - safely run slightly larger models with similar inference times to previously, or use larger context sizes with no real change in inference time, or just get a 10-15 second bump in generation time, all whilst keeping lore entries, instructions, and character cards. 

Originally all this stuff went at the start so that the model didn't get mixed up between instructions and style to copy. But there are ways around this. With context shifting I think there is motive to not have it there, at least optionally.

**Comment by e576082c** [comment]
I don't think this is the right way to do it. I may be wrong though. I'm just trying to summarize what I know about this. Maybe try to use the new memory field? Please have a look in the release notes:

**"New API feature: Split Memory - The generation payload also supports a new field memory in addition to the usual prompt field. If set, forcefully appends this string to the beginning of any submitted prompt text. If resulting context exceeds the limit, forcefully overwrites text from the beginning of the main prompt until it can fit. Useful to guarantee full memory insertion even when you cannot determine exact token count. Automatically used in Lite."**

**"NEW FEATURE: Context Shifting (A.K.A. EvenSmarterContext) - This feature utilizes KV cache shifting to automatically remove old tokens from context and add new ones without requiring any reprocessing. So long as you use no memory/fixed memory and don't use world info, you should be able to avoid almost all reprocessing between consecutive generations even at max context. This does not consume any additional context space, making it superior to SmartContext."**

I suppose, that this Context Shifting trick is maybe from [here](https://github.com/mit-han-lab/streaming-llm).
I also suppose, that using the Split Memory feature with this new feature is essential to make it work correctly. Please look at this comment of @LostRuins, from this [issue](https://github.com/LostRuins/koboldcpp/issues/445) (it's a long discussion!):
> "Submit memory and prompt separately so it'll be easier to match against."

**Edit:** I re-read what I just posted: **So long as you use no memory/fixed memory and don't use world info...**
Wait, what!? Uh, sorry I'm out of the loop, I just have read this [issue](https://github.com/LostRuins/koboldcpp/issues/523).

Another comment from @LostRuins:
> "Anyway, try the new version 1.49, the split memory may help"

**Comment by Cohee1207** [comment]
Context shifting should help you, that's true. Koboldcpp became very usable with this change. Also, I've recently added macro for char. card parts, so you can move them to the author's note and remove them from the story string if you wish.

**Comment by Drael64** [comment]
> **Edit:** I re-read what I just posted: **So long as you use no memory/fixed memory and don't use world info...** Wait, what!? Uh, sorry I'm out of the loop, I just have read this [issue](https://github.com/LostRuins/koboldcpp/issues/523).
> 
> Another comment from @LostRuins:
> 
> > "Anyway, try the new version 1.49, the split memory may help"

Yeah so basically the context shifting works like the old smart context - where if the previous chat history doesn't contact some changed or fixed component, that up to the point where you find such a change, it will re-use the context. So basically if all that character card, world info, prompt etc, isn't at the start, you get to re-used some context. Ofc, that presents the problem of how to phrase it, and which models to use that won't get confused by that. But it's potentially a way to save prompt time.

**Comment by Drael64** [comment]
> Context shifting should help you, that's true. Koboldcpp became very usable with this change. Also, I've recently added macro for char. card parts, so you can move them to the author's note and remove them from the story string if you wish.

Awesome sauce.! Thank you 🙏

 I don't know what the ideal instruct is to do it this way, or which model might handle this best, but being able to shift character cards, world info, everything normally at the start of the prompt to later in the prompt could be very powerful for getting more go out of ones hardware and worth having as an option!
 
 Cheers!

**Comment by trihardseven** [comment]
@Cohee1207 Can you elaborate on how to make context shifting work properly with SillyTavern? More often than not it just loads the full context for me, and when it does work I have no idea why. You said something about a macro, do I need to use that to make the work?

**Comment by Sazu-bit** [comment]
@Cohee1207 , I second trihardseven's request. I am using SillyTavern with KoboldCpp, I have no random insertions that I can find, lorebook is disabled... and yet still no ContextShifting. Likely to set up a new bug/help ticket thing but still.

**Comment by LostRuins** [comment]
You can observe the prompt being sent to KCPP in the terminal - check if there's anything that could prevent the subsequence from being correctly matched.
